Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -115,21 +124,53 @@ public struct HTTPRequestHead: Equatable {
     }
 }
 
-/// Internal representation of a URI
-enum URI {
-    case string(String)
-    case byteBuffer(ByteBuffer)
-}
+/// Represents a the URI component of an `HTTPRequestHead`. Allows the URI
+/// to be accessed as a `String` or as raw bytes.
+public struct HTTPURI: ExpressibleByStringLiteral {
+    /// Internal storage.
+    var storage: HTTPURIStorage
+
+    /// Internal init.
+    internal init(storage: HTTPURIStorage) {
+        self.storage = storage
+    }
+
+    /// See `ExpressibleByStringLiteral`.
+    public init(stringLiteral value: String) {
+        storage = .string(value)
+    }
 
-private extension String {
-    init(uri: URI) {
-        switch uri {
-        case .string(let string):
-            self = string
-        case .byteBuffer(let buffer):
-            self = buffer.getString(at: buffer.readerIndex, length: buffer.readableBytes)!
+    /// Access this `HTTPURI`'s raw bytes.
+    /// - Parameter closure: A closure accepting an `UnsafeRawBufferPointer` to this URI's raw bytes.
+    public func withUnsafeBytes<T>(_ closure: (UnsafeRawBufferPointer) -> T) -> T {
+        switch storage {
+        case .string(let s):
+            // cString + utf8 codeunit count -> raw buffer pointer
+            return s.withCString { closure(.init(start: .init($0), count: s.utf8.count)) }
+        case .byteBuffer(let b):
+            // point to readable byte region
+            return b.withUnsafeReadableBytes(closure)
         }
     }
+
+    /// Get or set the `String` representation of this `HTTPURI`.
+    public var string: String {
+        get {
+            switch storage {
+            case .string(let s): return s
+            case .byteBuffer(let b): return b.getString(at: b.readerIndex, length: b.readableBytes)!
+            }
+        }
+        set {
+            storage = .string(newValue)
+        }
+    }
+}
+
+/// Internal representation of a URI
+enum HTTPURIStorage {
```

Comment:
This guy can probably be an inner type on `HTTPURI`, which also means we can avoid the implicit name spacing in the type name.